pkg/scheduler/framework/plugins/topologyspreadconstraints/utils.go

```diff
@@ -6,15 +6,170 @@ Licensed under the MIT license.
+				//
+				// Assign a score even if the constraint does not concern the cluster.
+				scores[clusterName(cluster.Name)] += 0
+				continue
+			}
+
+			// The cluster under inspection is part of the spread.
+
+			// Verify if the placement will violate the constraint.
+			violated, skewChange, err := willViolate(domainCounter, domainName(val), int(*constraint.MaxSkew))
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to evaluate DoNotSchedule topology spread constraints: %w", err)
+			}
+			if violated {
+				// A violation happens.
+				reasons := violationReasons{
+					fmt.Sprintf(doNotScheduleConstraintViolationReasonTemplate, constraint.TopologyKey, *constraint.MaxSkew),
+				}
+				violations[clusterName(cluster.Name)] = reasons
+				continue
+			}
+			scores[clusterName(cluster.Name)] += skewChange * skewChangeScoreFactor
+		}
+	}
+
+	for _, constraint := range scheduleAnyway {
+		domainCounter := countByDomain(clusters, state, constraint.TopologyKey)
+
+		for _, cluster := range clusters {
+			if _, ok := violations[clusterName(cluster.Name)]; ok {
+				// The cluster has violated a DoNotSchedule topology spread constraint; no need
+				// evaluate other constraints for this cluster.
+				continue
+			}
+
+			val, ok := cluster.Labels[constraint.TopologyKey]
+			if !ok {
+				// The cluster under inspection does not have the topology key and thus is not part
+				// of the spread.
+				//
+				// Placing resources on such clusters will not lead to topology spread constraint
+				// violations.
+				//
+				// Assign a score even if the constraint does not concern the cluster.
+				scores[clusterName(cluster.Name)] += 0
+				continue
+			}
+
+			// The cluster under inspection is part of the spread.
+
+			// Verify if the placement will violate the constraint.
+			violated, skewChange, err := willViolate(domainCounter, domainName(val), int(*constraint.MaxSkew))
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to evaluate ScheduleAnyway topology spread constraints: %w", err)
+			}
+			if violated {
+				// A violation happens; since this is a ScheduleAnyway topology spread constraint,
+				// a violation score penality is applied to the score.
+				scores[clusterName(cluster.Name)] -= maxSkewViolationPenality
+				continue
+			}
+			scores[clusterName(cluster.Name)] += skewChange * skewChangeScoreFactor
+		}
+	}
+
+	return violations, scores, nil
+}
+
 // prepareTopologySpreadConstraintsPluginState initializes the state for the plugin to use
 // in the scheduling cycle.
-//
-// TO-DO (chenyu1): assign variables when needed.
-func prepareTopologySpreadConstraintsPluginState(_ framework.CycleStatePluginReadWriter, _ *fleetv1beta1.ClusterSchedulingPolicySnapshot) (*pluginState, error) {
-	// Not yet implemented.
-	return &pluginState{}, nil
+func prepareTopologySpreadConstraintsPluginState(state framework.CycleStatePluginReadWriter, policy *fleetv1beta1.ClusterSchedulingPolicySnapshot) (*pluginState, error) {
```

Comment:
just notice that the state is not used in this func

Reply:
Ah, line 163?

Reply:
ah, my bad